src/book.py

```diff
@@ -188,36 +188,95 @@ def _read_coinbase(self, file_path: Path) -> None:
 
                 # Unused variables.
                 del eur_total
-                del remark
 
                 # Validate data.
                 assert operation
                 assert coin
                 assert change
-                assert eur_spot
 
-                self.append_operation(
-                    operation, utc_time, platform, change, coin, row, file_path
-                )
+                if operation == "Convert":
+                    # parse change + coin from remark which is in format "0,123 ETH to 0,456 BTC"
+                    remark_target = remark.split(' to ')[-1]
+                    remark_target_parts = remark_target.split(' ')
+                    _convert_change = remark_target_parts[0].replace(',', '.')
+                    convert_change = misc.xdecimal(_convert_change)
+                    convert_coin = remark_target_parts[1]
+
+                    self.append_operation("Sell", utc_time, platform,
+                                          change, coin, row, file_path)
+                    self.append_operation("Buy", utc_time, platform,
+                                          convert_change, convert_coin, row, file_path)
+                else:
+                    self.append_operation(
+                        operation, utc_time, platform, change, coin, row, file_path
+                    )
+
+                    if operation == "Sell":
+                        assert isinstance(eur_subtotal, decimal.Decimal)
+                        self.append_operation(
+                            "Buy", utc_time, platform, eur_subtotal, "EUR", row, file_path
+                        )
+                    elif operation == "Buy":
+                        assert isinstance(eur_subtotal, decimal.Decimal)
+                        self.append_operation(
+                            "Sell", utc_time, platform, eur_subtotal, "EUR", row, file_path
+                        )
+
+                    if eur_fee:
```

Comment:
Is  this supposed to be indented?
Are `Convert` operations always without fees?

Reply:
There are no direct fees as Coinbase works with spread margins up to 2%: https://help.coinbase.com/en/coinbase/trading-and-funding/cryptocurrency-trading-pairs/convert-cryptocurrency-faq 

However I'll try to do some calculations again and check whether it's correct what I'm doing here. Especially since we got some information about the used prices in the CSV and have access to the Coinbase Pro price information. Perhaps a hybrid approach is necessary here.

Reply:
I've updated the Coinbase pricing mechanism by using the spot prices from the CSV again. See this sample:

```
Timestamp: 2019-04-03T08:51:36Z
Transaction Type: Convert
Asset: BTC
Quantity Transacted: 0.045368
EUR Spot Price at Transaction: 4396.49
EUR Subtotal: 384.43
EUR Total (inclusive of fees): 199.46
EUR Fees: 184.97
Notes: "Converted 0,045368 BTC to 0,8654877 BCH"
```

In our code this would be:
```python
coin = "BTC"
change = Decimal("0.045368")
convert_coint = "BCH"
convert_change = "0.8654877"
```

Which means that I can calculate the exchange rate of BCH by calculating `eur_total / convert_change` which is `199,45995832 EUR / 0,8654877 BCH = 230,459610598741 EUR/BCH`

A sanity check using the Coinbase Pro API: the BCH price at 2019-04-03 08:51:36 was 227.815 EUR/BCH

So although there are no fees, it's indeed important to use the values from the CSV as even Coinbase and Coinbase Pro prices differ and this is how they can allow converting without additionally charging fees.

Just out of curiosity because luckily it's irrelevant for the algorithm: if somebody finds out what the values of Coinbase's columns "EUR Subtotal" and "EUR Fees" are, please let me know. I know for sure that I didn't pay 184.97 EUR of fees 😊